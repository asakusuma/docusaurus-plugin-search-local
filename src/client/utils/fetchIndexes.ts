import lunr from "lunr";
import { SearchDocument, SearchDocumentType, WrappedIndex } from "../../types";

interface SerializedIndex {
  documents: SearchDocument[];
  index: {
    invertedIndex: [string][];
  };
}

export async function fetchIndexes(
  baseUrl: string,
  indexHash: string | null = null
): Promise<{
  wrappedIndexes: WrappedIndex[];
  zhDictionary: string[];
}> {
  const indexUrl = `${baseUrl}search-index.json`;
  const queryString = indexHash ? `?_=${indexHash}` : "";

  const json = (await (
    await fetch(`${indexUrl}${queryString}`)
  ).json()) as SerializedIndex[];

  const wrappedIndexes: WrappedIndex[] = json.map(
    ({ documents, index }, type) => ({
      type: type as SearchDocumentType,
      documents,
      index: lunr.Index.load(index),
    })
  );

  return {
    wrappedIndexes,
    zhDictionary: [], // TODO: strip this out
  };
}
