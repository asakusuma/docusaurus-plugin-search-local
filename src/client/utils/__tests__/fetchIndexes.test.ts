import lunr from "lunr";
import { fetchIndexes } from "../fetchIndexes";

jest.mock("lunr");

const mockLunrIndexLoad = (
  jest.spyOn(lunr.Index, "load") as any
).mockImplementation(() => `loaded-index`);

const mockFetch = (global.fetch = jest.fn());

describe("fetchIndexes", () => {
  const baseUrl = "/";

  beforeEach(() => {
    jest.resetModules(); // most important - it clears the cache
    mockFetch.mockReset();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  test("production with empty index", async () => {
    process.env.NODE_ENV = "production";
    mockFetch.mockResolvedValueOnce({
      json: () => Promise.resolve([]),
    });
    const result = await fetchIndexes(baseUrl, "abc");
    expect(mockFetch).toBeCalledWith("/search-index.json?_=abc");
    expect(result).toEqual({
      wrappedIndexes: [],
      zhDictionary: [],
    });
  });

  test("production", async () => {
    process.env.NODE_ENV = "production";
    mockLunrIndexLoad;
    mockFetch.mockResolvedValueOnce({
      json: () =>
        Promise.resolve([
          {
            documents: [1, 2, 3],
            index: {
              invertedIndex: [["hello"], ["hello"]],
            },
          },
        ]),
    });
    const result = await fetchIndexes(baseUrl, "abc");
    expect(mockFetch).toBeCalledWith("/search-index.json?_=abc");
    expect(result).toEqual({
      wrappedIndexes: [
        {
          documents: [1, 2, 3],
          index: "loaded-index",
          type: 0,
        },
      ],
      zhDictionary: [],
    });
  });

  test("it should only add the index to the URL if provided", async () => {
    mockFetch.mockResolvedValue({
      json: () => Promise.resolve([]),
    });

    await fetchIndexes(baseUrl, "abc");

    // Standard call with a hash
    expect(mockFetch).toBeCalledWith("/search-index.json?_=abc");

    // Not providing a hash index
    await fetchIndexes(baseUrl);

    // Should not add query param for hash index if generated hash is `null`.
    expect(mockFetch).toBeCalledWith("/search-index.json");
  });
});
