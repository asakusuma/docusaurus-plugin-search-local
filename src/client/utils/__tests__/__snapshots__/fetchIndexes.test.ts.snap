// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchIndexes it should handle endpoint errors 1`] = `
Object {
  "wrappedIndexes": Array [],
}
`;

exports[`fetchIndexes it should handle errors parsing the search index file 1`] = `
Object {
  "wrappedIndexes": Array [],
}
`;

exports[`fetchIndexes production 1`] = `
Object {
  "wrappedIndexes": Array [
    Object {
      "documents": Array [
        Object {
          "h": "#installation",
          "i": 12,
          "p": 10,
          "t": "Installation",
          "u": "/docusaurus-plugin-search-local/docs/getting-started/",
        },
      ],
      "index": Object {
        "fieldVectors": Array [
          Array [
            "t/12",
            Array [
              0,
              0.981,
            ],
          ],
          Array [
            "t/14",
            Array [
              1,
              0.981,
            ],
          ],
          Array [
            "t/16",
            Array [
              2,
              0.981,
            ],
          ],
        ],
        "fields": Array [
          "t",
        ],
        "invertedIndex": Array [
          Array [
            "hello",
            Object {
              "_index": 1,
              "t": Object {
                "14": Object {
                  "position": Array [
                    Array [
                      0,
                      13,
                    ],
                  ],
                },
              },
            },
          ],
          Array [
            "world",
            Object {
              "_index": 0,
              "t": Object {
                "12": Object {
                  "position": Array [
                    Array [
                      0,
                      12,
                    ],
                  ],
                },
              },
            },
          ],
        ],
        "pipeline": Array [
          "stemmer",
        ],
        "version": "2.3.9",
      },
      "type": 0,
    },
  ],
}
`;

exports[`fetchIndexes production with empty index 1`] = `
Object {
  "wrappedIndexes": Array [],
}
`;
